libs/parallelism/algorithms/include/hpx/parallel/container_algorithms/rotate.hpp

```diff
@@ -122,26 +363,291 @@ namespace hpx { namespace parallel { inline namespace v1 {
     /// within each thread.
     ///
     /// \returns  The \a rotate_copy algorithm returns a
-    ///           \a hpx::future<tagged_pair<tag::in(FwdIter), tag::out(OutIter)> >
+    ///           \a hpx::future<otate_copy_result<
+    ///           hpx::traits::range_iterator_t<Rng>, OutIter>>
     ///           if the execution policy is of type
     ///           \a parallel_task_policy and
-    ///           returns \a tagged_pair<tag::in(FwdIter), tag::out(OutIter)>
+    ///           returns \a rotate_copy_result<
+    ///           hpx::traits::range_iterator_t<Rng>, OutIter>
     ///           otherwise.
-    ///           The \a rotate_copy algorithm returns the output iterator to the
-    ///           element past the last element copied.
+    ///           The \a rotate_copy algorithm returns the output iterator to
+    ///           the element past the last element copied.
     ///
+    template <typename ExPolicy, typename Rng, typename OutIter>
+    typename parallel::util::detail::algorithm_result<ExPolicy,
+        rotate_copy_result<hpx::traits::range_iterator_t<Rng>, OutIter>>::type
+    rotate_copy(ExPolicy&& policy, Rng&& rng,
+        hpx::traits::range_iterator_t<Rng> middle, OutIter dest_first);
+
+}}    // namespace hpx::ranges
+
+#else
+
+#include <hpx/config.hpp>
+#include <hpx/concepts/concepts.hpp>
+#include <hpx/iterator_support/range.hpp>
+#include <hpx/iterator_support/traits/is_iterator.hpp>
+#include <hpx/iterator_support/traits/is_range.hpp>
+
+#include <hpx/algorithms/traits/projected_range.hpp>
+#include <hpx/iterator_support/iterator_range.hpp>
+#include <hpx/parallel/algorithms/rotate.hpp>
+#include <hpx/parallel/tagspec.hpp>
+#include <hpx/parallel/util/projection_identity.hpp>
+#include <hpx/parallel/util/result_types.hpp>
+
+#include <type_traits>
+#include <utility>
+
+namespace hpx { namespace parallel { inline namespace v1 {
+
+    // clang-format off
+    template <typename ExPolicy, typename Rng,
+        HPX_CONCEPT_REQUIRES_(
+            hpx::is_execution_policy<ExPolicy>::value &&
+            hpx::traits::is_range<Rng>::value
+        )>
```

Comment:
`hpx::is_execution_policy_v<ExPolicy>` is defined here: https://github.com/STEllAR-GROUP/hpx/blob/master/libs/parallelism/execution/include/hpx/execution/traits/is_execution_policy.hpp#L67-L69, we still miss a `hpx::traits::is_range_v<Rng>` however (see: https://github.com/STEllAR-GROUP/hpx/blob/master/libs/core/iterator_support/include/hpx/iterator_support/traits/is_range.hpp)

Reply:
ok, I see.  I will check and see if I can add a `hpx::traits::is_range_v<Rng>` for `hpx::traits::is_range<Rng>::value` later, I mean as another minor task. if you think this is necessary for our future work.

Reply:
We have a lot of similar things to work on. This is nothing for this PR, though.